### PR TITLE
[fix] Rewritten the algorithm used to get the reference position

### LINKED ISSUE #171: placement: 'top' - does not flip the popper when it touches the upper viewport boundary
### Steps to reproduce the problem
1.  Visit https://popper.js.org
2.  In the first example (#example10positionSelector) choose 'top'
3.  Scroll down

### What is the expected behavior?
Popper should be placed under the trigger, as soon as the upper viewport boundary is hitting the popper's top border.

### What went wrong?
Popper is not flipped to bottom.

### Any other comments?
Choosing placement: "bottom" works as expected

### LINKED ISSUE #188: Positioning incorrect when popper inside relatively or absolutely positioned parent
### CodePen demo

http://codepen.io/anon/pen/gmRRwz

### Steps to reproduce the problem

1. Instantiate a new popper, with default settings, inside of a relatively or absolutely positioned  parent.
2. Scroll or resize the page so that popper hits the bottom of the viewport, thus causing it to flip to the other side.

### What is the expected behavior?

Initial placement should be correct.

screen shot 2017-03-13 at 8 18 46 am

The popper should flip to the opposite side of the target.

screen shot 2017-03-13 at 8 14 14 am

### What went wrong?

The initial placement is incorrect.

screen shot 2017-03-13 at 8 17 47 am

The popper flips, but the placement is incorrect.

screen shot 2017-03-13 at 8 14 28 am

### Any other comments?

Not at this time.

### PR DESCRIPTION
**Pull request by FezVrasta**
This work is needed to:

1. fix #188
1. fix #171
1. Probably #198
1. (bonus) reduce size
1. (bonus) reduce complexity

Some/lot of tests are failing, any help to fix them is super welcome (really!). Just fork this branch, apply your changes, and send a PR against this branch.

Thank you.

### CODE REVIEW

File: src/popper/utils/findCommonOffsetParent.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import isOffsetContainer from './isOffsetContainer';
+
+export default function findCommonOffsetParent(element1, element2) {
```
Comment (FezVrasta) [inline src/popper/utils/findCommonOffsetParent.js]: One of the Popper problems was that we didn't consider that reference and popper could live in two different offset contexts.

Now we find the common offset parent and use it as reference to compute all the positions.

File: src/popper/utils/getOffsetRectRelativeToArbitraryNode.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import getStyleComputedProperty from './getStyleComputedProperty';
+
+export default function getOffsetRectRelativeToArbitraryNode(children, parent) {
```
Comment (FezVrasta) [inline src/popper/utils/getOffsetRectRelativeToArbitraryNode.js]: This is basically an improved version of `getOffsetRectRelativeToCustomParent`

**Comment by FezVrasta** [comment]
There are only 3 tests failing in the `core` suite and few failing in `Tooltip.js`.

The ones in `core` are about Shadow DOM and they are related to the new way I use to find the common offset parent, I think it should be quite easy to workaround tho.

I had no time to look at the Tooltip.js tests, but I don't expect them to be broken in bad ways...

File: tests/functional/flips.js [hunk 3]
```diff
@@ -166,13 +166,11 @@ describe('[flipping]', () => {
 
     xit('flips to opposite side when rendered inside a positioned parent', (done) => {
```
Comment (nadiam84) [inline tests/functional/flips.js]: Should this be un-xit also? :)

File: src/popper/modifiers/preventOverflow.js [hunk 4]
```diff
@@ -24,8 +24,9 @@ import getBoundaries from '../utils/getBoundaries';
  */
 export default function preventOverflow(data, options) {
     const boundariesElement = options.boundariesElement || getOffsetParent(data.instance.popper);
-    const boundaries = getBoundaries(data.instance.popper, options.padding, boundariesElement);
+    const boundaries = getBoundaries(data.instance.popper, data.instance.reference, options.padding, boundariesElement);
     options.boundaries = boundaries;
+    console.log('boundaries', {...boundaries});
```
Comment (nadiam84) [inline src/popper/modifiers/preventOverflow.js]: Should it be here? :)

Comment (FezVrasta) [inline src/popper/modifiers/preventOverflow.js]: It's still under development, I'll clean everything later

File: tests/functional/flips.js (hunk 3 above)
Comment (FezVrasta) [inline tests/functional/flips.js]: I'm going to work on these once I get the existing tests passing

**Comment by FezVrasta** [comment]
Fixed the tooltips tests, now we have only the 3 shadow DOM tests failing (wuut!)

**Comment by FezVrasta** [comment]
So, 1 test is failing on Firefox and Chrome on CI (no idea why, locally they work...) and some are failing on IE and Edge. Will look at them tomorrow. Feel free to pick up the branch and fix them if you feel in mood. (I'll stop asking it at some point...)

**Comment by FezVrasta** [comment]
The failing test on Chrome and Firefox has been fixed (even if not really cleanly actually...), now I'm working on the latest `x`'ed test.

**Comment by FezVrasta** [comment]
And #188 is fixed as well.

img

Now, it's your turn Internet Explorer.

**Comment by FezVrasta** [comment]
So, IE11 and Edge now pass. But IE10 is hard to die.

I can't work on it anymore for today, again, if someone (maybe someone that uses this library in its multi billionaire company, Microsoft, anyone?) could spend some time and fix it for me. 🙂

**Comment by FezVrasta** [comment]
Seems like this PR introduced a regression, note that it doesn't take margins of `body` into account:
https://codepen.io/FezVrasta/pen/qmEObw?editors=1100

**Comment by FezVrasta** [comment]
I thought I was screwed because I used the "party hard" gif in a previous message, but luckily now that **all our test suite pass** I found this one:

img

**Pull request merged**